eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpCommand.java

```diff
@@ -54,12 +58,17 @@ void warp(@Context Player player, @Context User user) {
     @Execute(name = "warp")
     @DescriptionDocs(description = "Teleport to warp, if player has permission eternalcore.warp.bypass teleport will be instant", arguments = "<warp>")
     void warp(@Context Player player, @Arg Warp warp) {
-        if (player.hasPermission("eternalcore.warp.bypass")) {
-            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), warp.getPosition(), Duration.ZERO);
+        WarpTeleportEvent event = this.eventCaller.callEvent(new WarpTeleportEvent(player, warp));
+        if (event.isCancelled()) {
             return;
         }
 
-        this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), warp.getPosition(), Duration.ofSeconds(5));
+        if (player.hasPermission("eternalcore.warp.bypass")) {
+            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), PositionAdapter.convert(warp.getLocation()), Duration.ZERO);
+        }
+        else {
```

Comment:
no else needed, just remove it in permission check create return and will be great